tests/fuzzers/nodestatemachine/nodestatemachine-fuzzer.go

```diff
@@ -0,0 +1,479 @@
+	f.setup = s
+	ns := nodestate.NewNodeStateMachine(mdb, []byte("-ns"), clock, s)
+	var callbackCount int
+	var ops [256]func(n *enode.Node, sub bool)
+
+	runsub := func(n *enode.Node, op byte) {
+		if fn := ops[op]; fn != nil {
+			fn(n, true)
+		}
+	}
+
+	var allFlags nodestate.Flags
+	for _, f := range flags {
+		allFlags = allFlags.Or(f)
+	}
+	var offState state
+
+	var flagSubs [16]struct {
+		a, b, c int
+		ops     [8]byte
+	}
+	shift := byte(42)
+	for i := range flagSubs {
+		x := f.randomInt(flagCount * flagCount * flagCount)
+		flagSubs[i].a = x % flagCount
+		x /= flagCount
+		flagSubs[i].b = x % flagCount
+		x /= flagCount
+		flagSubs[i].c = x % flagCount
+		var subops [8]byte
+		for j := range subops {
+			flagSubs[i].ops[j] = f.randomByte() + shift
+			shift += 77
+		}
+	}
+
+	var fieldSubs [fieldCount]struct {
+		ops [8]byte
+	}
+	for ii := range fields {
+		var subops [8]byte
+		for j := range subops {
+			fieldSubs[ii].ops[j] = f.randomByte() + shift
+			shift += 77
+		}
+	}
+
+	addSubs := func() {
+		ns.SubscribeState(allFlags.Or(s.OfflineFlag()), func(n *enode.Node, oldState, newState nodestate.Flags) {
+			o1, o2 := oldState.HasAll(s.OfflineFlag()), newState.HasAll(s.OfflineFlag())
+			if o1 != o2 {
+				var st nodestate.Flags
+				if o1 {
+					st = newState
+				} else {
+					st = oldState
+				}
+				var m uint64
+				for _, f := range flags {
+					m += m
+					if st.HasAll(f) {
+						m++
+					}
+				}
+				offState[int(n.ID()[0])].mask = m
+			}
+		})
+
+		for _, sub := range flagSubs {
+			ns.SubscribeState(nodestate.MergeFlags(s.OfflineFlag(), flags[sub.a], flags[sub.b], flags[sub.c]), func(n *enode.Node, oldState, newState nodestate.Flags) {
+				if oldState.Or(newState).HasAll(s.OfflineFlag()) {
+					return
+				}
+				callbackCount++
+				if callbackCount > cbLimit {
+					return
+				}
+				j := 0
+				if newState.HasAll(flags[sub.a]) {
+					j += 1
+				}
+				if newState.HasAll(flags[sub.b]) {
+					j += 2
+				}
+				if newState.HasAll(flags[sub.c]) {
+					j += 4
+				}
+				runsub(n, sub.ops[j])
+			})
+		}
+
+		for ii, fi := range fields {
+			fieldIdx := ii
+			ns.SubscribeField(fi, func(n *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+				if state.HasAll(s.OfflineFlag()) {
+					v := oldValue
+					if v == nil {
+						v = newValue
+					}
+					var vv uint16
+					if uv, ok := v.(uint64); ok {
+						vv = 256 + uint16(byte(uv))
+					}
+					if uv, ok := v.(uint16); ok {
+						vv = 512 + uint16(byte(uv))
+					}
+					offState[int(n.ID()[0])].fields[fieldIdx] = vv
+					return
+				}
+
+				callbackCount++
+				if callbackCount > cbLimit {
+					return
+				}
+				var j int
+				if newValue != nil {
+					if u, ok := newValue.(uint64); ok {
+						j = int(u%7) + 1
+					} else {
+						j = int(u%7) + 1
+					}
+				}
+				runsub(n, fieldSubs[fieldIdx].ops[j])
+			})
+		}
+	}
+	for i := range ops {
+		b := f.randomByte()
+		if b+byte(i*137) < 4 {
```

Comment:
What's the magic of `137`?

Reply:
The only magic here is that it is an odd number and therefore a relative prime to 256 :) So the result of the modulo 256 multiplication sweeps the entire byte range and `< 4` can work as a simple quasi-random condition. What I wanted here is to make the set of single operations variable in size so if the first byte of the pair satisfies a quasi-random condition then it is considered the end of the set. The chance of satisfying the condition is 1/64 so the average size of the set is 64.